LINDA Reforged

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Much thanks to duncathan, Aranclanos, and skog for their help with understanding the atmos module, and their advice and review of the documentation included in this pr.

Speeds up gas movement significantly
Documents the intent and finer details of the atmos system (Thanks dunc)
Fixes excited groups constantly rebuilding, this broke 4 years ago
Fixes superconductors just straight up not working
Allows turfs to sleep while inside an excited group
Adds a new subprocess to SSAir to support rebuilding in this state
Most heat based behavior no longer relies on being inside a fire
Adds a new element to support doing this cleanly
Adds a new subprocess to SSAir to support doing this while a turf is asleep
Refactors air_update_turf to allow for finer control
Makes apcs take damage in heat to prevent infinite plasma fire diffs
Cleans up immutable gas mixtures to make them work properly when the mix has gas in it
Planetary turfs no longer create a new copy of themselves each time they process. We instead use a global
immutable mix
Cleans up a typed for loop in reactions
Canisters will take damage from outside heat now
Speeds up excited group dismantle
Increases the superconductor threshold by 200k
Cleans up some roundstart ATs on some ruins
Uses /turf/open/var/excited to track if a turf is actively processing, preventing a |=
Prevents openspace from trying to melt
Tweaks a canister examine line
Makes planetary turfs reset to base when broken down as part of an excited group
Makes it impossible for planetary turfs to rebuild, just like space tiles
Fixes closed turfs not activating their replacement when destroyed by moving closed -> open turf activation to the adjacent air subsystem. They were activating and then going back to sleep before adjacent air got a chance to tick.
Fire alarms will trigger when the area gets too cold for humans

Fixes #45448
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Atmos should actually work as intended.
(I know auxmos is looming, but I'd like our atmospherics system to actually work so we have a fallback if something breaks)
Most of this pr is grunt work, converting heat reactions/air_update_turf over, for details on how environmental atmos should work and why I did what I did, please check the documentation.

This pr brings some accidental balance changes purely as a result of some old dusty atmos feature actually working again.
Superconduction actually working in particular will be a large change, but I think it's one that's worth going through.

[The atmospherics documentation](https://github.com/tgstation/tgstation/blob/2bdedc90f402d901cc42a2b9bd093d9932fe7867/code/modules/atmospherics/Atmospherics.md) contains some external images, currently hosted on a local branch of mine. I was planning on pr'ing them to a separate repo in the tg org, but I'd like your thoughts on that.

I'm sorry if I've missed a change by the way, this should be the brunt of it.

See https://github.com/LemonInTheDark/tgstation/tree/sleep and #52493 for the full commit history and a more detailed explanation of the excited group issue, I thought it was just a bit too long.


## Changelog
:cl:
fix: Speeds up gas movement
add: Documents the intent and finer details of the atmos system (Thanks dunc/aran/skog)
fix: Fixes excited groups constantly rebuilding, this broke 4 years ago
fix: Fixes superconduction just straight up not working. This means heat can leak through windows, glass doors, and a few other things. Keep an eye on your temperature
fix: Most heat based behavior no longer relies on being inside a fire
add: APCs take damage in heat to prevent infinite plasma fire diffs, watch out when operating that supermatter
refactor: Cleans up immutable gas mixtures to make them work properly when the mix has gas in it
refactor: Planetary turfs no longer create a new copy of themselves each time they process. We instead use a global
mmutable mix
refactor: Cleans up a typed for loop in reactions
fix: Canisters will properly take damage from outside heat now
fix: Cleans up some roundstart ATs on some ruins
fix: Keeps openspace from trying to melt
fix: Destroying a wall wakes up the tile beneath it to atmos, no more meteor breaches with no flow
add: Fire alarms will trigger when the area gets too cold for humans
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros commented:
It appears that you too have been struck by linting comments... Why are we linting comments?

LemonInTheDark commented:
Linting comment syntax is fine, I really did just mess up

Ghilker commented:
Yes! Finally he's back!

AffectedArc07 commented:
![image](https://user-images.githubusercontent.com/25063394/102622613-2e42c080-4139-11eb-99b8-bb11d4928fdb.png)
Your CL tags are backwards

Qustinnus commented:
auxmos is "looming" in the sense that we might see it somewhere in a year or so

Qustinnus commented:
auxmos is "looming" in the sense that we might see it somewhere in a year or so

Qustinnus commented on code/__DEFINES/dcs/signals.dm:
```diff
@@ -872,6 +872,10 @@
 ///from monkey CtrlClickOn(): (/mob)
 #define COMSIG_XENO_MONKEY_CLICK_CTRL "xeno_monkey_click_ctrl"
 
+// /datum/component/atmos_sensitive
```
its an element, fix the comments pls (I was looking for a component and couldnt find it)

Qustinnus replied:
actually why does it mention the element anyways? The signal is not linked to the element in any way and comments shouldn't give the idea that they're made for specific things; they just represent the action of temp exposure happening.

Qustinnus replied:
actually why does it mention the element anyways? The signal is not linked to the element in any way and comments shouldn't give the idea that they're made for specific things; they just represent the action of temp exposure happening.

LemonInTheDark replied:
local man refactors his heat sensitivity system 3 times, forgets how it works when writing documentation

Qustinnus replied:
its ok old man ur almost there

Qustinnus commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,66 @@
+		SSair.atom_process -= src
+		flags_1 &= ~ATMOS_IS_PROCESSING_1
+
+/atom/proc/process_exposure()
```
This proc looks like it could use some OOP, unless you think that would be worse off for performance? I can't imagine that doing istype on src is ever a good thing though.

LemonInTheDark replied:
I've updated it, maybe not in the best way but it's more clean now.

Qustinnus commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,66 @@
+		SSair.atom_process -= src
+		flags_1 &= ~ATMOS_IS_PROCESSING_1
+
+/atom/proc/process_exposure()
```
This proc looks like it could use some OOP, unless you think that would be worse off for performance? I can't imagine that doing istype on src is ever a good thing though.

Qustinnus commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,66 @@
+
+/datum/element/atmos_sensitive/proc/handle_move(datum/source, atom/movable/oldloc, direction, forced)
+	var/atom/microchipped_lad = source
+	if(!istype(microchipped_lad))
```
there is no point checking this. There's no posibility of the source not being an atom; if it is there's a SERIOUS problem and it should cause a runtime. You check isatom before you register for this signal, so you know it's an atom already. Just casting is enough.

LemonInTheDark replied:
mood

Qustinnus commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,66 @@
+
+/datum/element/atmos_sensitive/proc/handle_move(datum/source, atom/movable/oldloc, direction, forced)
+	var/atom/microchipped_lad = source
+	if(!istype(microchipped_lad))
```
there is no point checking this. There's no posibility of the source not being an atom; if it is there's a SERIOUS problem and it should cause a runtime. You check isatom before you register for this signal, so you know it's an atom already. Just casting is enough.

Qustinnus commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,66 @@
+	if(!istype(microchipped_lad))
+		return
+	microchipped_lad.UnregisterSignal(oldloc, COMSIG_TURF_EXPOSE)
+	if(istype(microchipped_lad.loc, /turf/open))
```
weird thought maybe but all of these checks for open turf make me think we should just add a turf flag for being open / atmos check worthy.

LemonInTheDark replied:
it's been talked about in the past, it'd be a pretty big refactor, and I think it would make some code even more unreadable

Qustinnus replied:
is there such a big difference between (microchipped_lad.turf_flags & CAN_CONTAIN_ATMOS)  and this?

Qustinnus replied:
readability wise

LemonInTheDark replied:
I don't mean here, I mean with proc overrides and such, changeTurf() is a good example.

Qustinnus commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,66 @@
+	if(!istype(microchipped_lad))
+		return
+	microchipped_lad.UnregisterSignal(oldloc, COMSIG_TURF_EXPOSE)
+	if(istype(microchipped_lad.loc, /turf/open))
```
weird thought maybe but all of these checks for open turf make me think we should just add a turf flag for being open / atmos check worthy.

Qustinnus reviewed (changes requested):
some small comments

Cyberboss commented:
Based af

JJRcop commented:
![flame_of_the_west](https://user-images.githubusercontent.com/5211576/102646412-fe43ef00-4131-11eb-9a00-9c1fda41825c.gif)

LemonInTheDark commented:
> 
> 
> ![image](https://user-images.githubusercontent.com/25063394/102622613-2e42c080-4139-11eb-99b8-bb11d4928fdb.png)
> Your CL tags are backwards

Help I'm dumb

LemonInTheDark commented:
If you're seeing this, go read the documentation, no point writing it down if I'm the only one who ever reads it

Dennok commented:
Great!

LemonInTheDark commented:
I found another atmos bug, because #https://github.com/tgstation/tgstation/pull/43447 made closed turfs update their adjacent members after a bit of a wait, a closed turf could become active, then die before atmos_adjacent_turfs updated. I've fixed this by integrating closed turf activation with the adjacent_air ss, in only a slightly hacky way.

Cyprex commented:
Is this ready for a testmerge?

LemonInTheDark commented:
Always has been

Timberpoes commented:
I think this causes the meta kitchen freezer to be 24/7 alarm mode.

I've also seen Icebox in 24/7 station-wide fire alarm mode with this testmerged.

Both due to I assume cold temperatures. Cannot replicate on local.

LemonInTheDark commented:
Fire alarms triggering when cold is intended, the code for it was always there actually, it only checked when there was a fire tho, so it never triggered. I'll need to make a change to meta alongside this, but I'm delaying that because I don't want to enter another conflict hell

redguy999 commented:
I believe fire alarms may be bugged because of this testmerge, fire alarms seem to retrigger after being manually untriggered, even if it is being deconstructed, or if the temperature returns to normal.

LemonInTheDark commented:
> I believe fire alarms may be bugged because of this testmerge, fire alarms seem to retrigger after being manually untriggered, even if it is being deconstructed, or if the temperature returns to normal.

Couple things to note, the behavior around fire alarms and triggering/being triggered while in the process of deconning is the same as it's always been, it's based on the area itself, in a rather jank way. I agree that it needs worked on, but it isn't an issue with this pr in particular, and while allowing cold things to trigger it is making this issue more apparent, I'm doing my best to not feature creep this pr any further. If the other maintainers would like me to take a crack at it here I'd be willing to.

Retriggering when the area has returned to normal would be a bug, but I can't seem to reproduce it, do you have any details? Keep in mind, just because one fire alarm is off doesn't mean they all are, it works based on area rather then alarm by alarm.

AnturK commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -0,0 +1,70 @@
+	if(!isatom(target)) //How
+		return ELEMENT_INCOMPATIBLE
+	var/atom/to_track = target
+	to_track.RegisterSignal(get_turf(to_track), COMSIG_TURF_EXPOSE, /atom/proc/check_atmos_process)
```
Shouldn't this be loc, if something starts in an object, after next move Unregister won't remove the old signal handler.

LemonInTheDark replied:
Ah yeah you're right, I'll add a check for turfs here.

AnturK commented on code/game/turfs/change_turf.dm:
```diff
@@ -273,13 +282,14 @@ GLOBAL_LIST_INIT(blacklisted_automated_baseturfs, typecacheof(list(
 
 //If you modify this function, ensure it works correctly with lateloaded map templates.
-/turf/proc/AfterChange(flags) //called after a turf has been replaced in ChangeTurf()
+/turf/proc/AfterChange(flags, oldType, screaming) //called after a turf has been replaced in ChangeTurf()
```
screaming yes

LemonInTheDark replied:
woop

AnturK reviewed (changes requested):
Also conflicts

LemonInTheDark committed:
LINDA Reforged

Speeds up gas movement significantly
Documents the intent and finer details of the atmos system (Thanks dunc)
Fixes excited groups constantly rebuilding, this broke 4 years ago
Fixes superconductors just straight up not working
Allows turfs to sleep while inside an excited group
Adds a new subprocess to SSAir to support rebuilding in this state
Most heat based behavior no longer relies on being inside a fire
Adds a new element to support doing this cleanly
Adds a new subprocess to SSAir to support doing this while a turf is asleep
Refactors air_update_turf to allow for finer control
Makes apcs take damage in heat to prevent infinite plasma fire diffs
Cleans up immutable gas mixtures to make them work properly when the mix has gas in it
Planetary turfs no longer create a new copy of themselves each time they process. We instead use a global
immutable mix
Cleans up a typed for loop in reactions
Canisters will take damage from outside heat now
Speeds up excited group dismantle
Increases the superconductor threshold by 200k
Cleans up some roundstart ATs on some ruins
Uses /turf/open/var/excited to track if a turf is actively processing, preventing a |=
Prevents openspace from trying to melt
Tweaks a canister examine line
Makes planetary turfs reset to base when broken down as part of an excited group
Makes it impossible for planetary turfs to rebuild, just like space tiles
Fixes closed turfs not activating their replacement when destroyed by moving closed -> open turf activation to
the adjacent air subsystem. They were activating and then going back to sleep before adjacent air got a chance
to tick.
Fire alarms will trigger when the area gets too cold for humans

LemonInTheDark commented:
We have entered the endgame

ShibaInuLord commented:
pog?

Cyberboss commented:
Bless this PR

Putnam3145 commented:
based

> I know auxmos is looming

if this works it might be better to use this just to avoid funny rust overhead and the bus factor that is my own person
